File: oolab/src/main/java/agh/ics/oop/model/MapDirection.java
package agh.ics.oop.model;

public enum MapDirection {
    NORTH(new Vector2D(0,1),"Północ"),
    EAST(new Vector2D(1,0),"Wschód"),
    SOUTH(new Vector2D(0,-1),"Południe"),
    WEST(new Vector2D(-1,0),"Zachód");

    private Vector2D directionVector;
    private String directionName;

    MapDirection(Vector2D directionVector, String directionName) {
        this.directionVector = directionVector;
        this.directionName = directionName;
    }


    public MapDirection next() {
        return switch(this) {
            case NORTH -> EAST;
            case EAST -> SOUTH;
            case SOUTH -> WEST;
            case WEST -> NORTH;
        };
    }

    public MapDirection previous() {
        return switch(this) {
            case NORTH -> WEST;
            case EAST -> NORTH;
            case SOUTH -> EAST;
            case WEST -> SOUTH;
        };
    }

    public Vector2D toUnitVector() {
        return directionVector;
    }

    public String toString() {
        return directionName;
    }
}

File: kotlinlab/src/main/kotlin/model/MapDirection.kt
package model

enum class MapDirection(internal val directionVector: Vector2D, private val directionName : String) {
    NORTH(Vector2D(0,1),"Północ"),
    EAST(Vector2D(1,0),"Wschód"),
    SOUTH( Vector2D(0,-1),"Południe"),
    WEST(Vector2D(-1,0),"Zachód");

    fun next(): MapDirection {
        return when(this) {
            NORTH -> EAST
            EAST -> SOUTH
            SOUTH -> WEST
            WEST -> NORTH
        }
    }

    fun previous(): MapDirection {
        return when(this) {
            NORTH -> WEST
            EAST -> NORTH
            SOUTH -> EAST
            WEST -> SOUTH
        }
    }

    override fun toString() = directionName
}